mlflow/pyfunc/__init__.py

```diff
@@ -828,13 +828,185 @@ def _create_model_downloading_tmp_dir(should_use_nfs):
+    """
+    Convert list or numpy array values to spark dataframe column values.
+    """
+    from pyspark.sql.types import ArrayType
+
+    # Get element type from a array<E> type or a array<array<E>> type.
+    # and get the array_dim value (1 or 2).
+    array_dim = 1
+    elem_type = spark_type.elementType
+    if isinstance(elem_type, ArrayType):
+        elem_type = elem_type.elementType
+        array_dim = 2
+
+    if type(elem_type) not in spark_primitive_type_to_np_type:
+        raise MlflowException(
+            "Unsupported array type field with element type "
+            f"{elem_type.simpleString()} in struct type.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+    np_type = spark_primitive_type_to_np_type[type(elem_type)]
+
+    if array_dim == 1:
+        return [np.array(v, dtype=np_type) for v in values]
+    else:
+        return [list(np.array(v, dtype=np_type)) for v in values]
+
+
+def _get_spark_primitive_type():
+    from pyspark.sql.types import (
+        IntegerType,
+        LongType,
+        FloatType,
+        DoubleType,
+        StringType,
+        BooleanType,
+    )
+    return (
+        IntegerType,
+        LongType,
+        FloatType,
+        DoubleType,
+        StringType,
+        BooleanType,
+    )
+
+
+def _check_udf_return_struct_type(struct_type):
+    from pyspark.sql.types import StructType, ArrayType
+
+    primitive_types = _get_spark_primitive_type()
+
+    for field in struct_type.fields:
+        field_type = field.dataType
+
+        if isinstance(field_type, primitive_types):
+            pass
+        elif isinstance(field_type, ArrayType):
+            _check_udf_return_array_type(field_type, allow_nested_struct=False)
+        else:
+            raise MlflowException(
+                f"'spark_udf' return type does not support field type {field_type} "
+                "inside a struct type",
+                error_code=INVALID_PARAMETER_VALUE,
+            )
+
+
+def _check_udf_return_array_type(array_type, allow_nested_struct):
+    from pyspark.sql.types import ArrayType, StructType
+
+    elem_type = array_type.elementType
+    primitive_types = _get_spark_primitive_type()
+
+    if isinstance(elem_type, primitive_types):
+        # 1D array
+        return
+
+    if isinstance(elem_type, StructType):
+        if allow_nested_struct:
+            # Array of struct values.
+            _check_udf_return_struct_type(elem_type)
+            return
+
+        raise MlflowException(
+            f"'spark_udf' return type does not support struct type nesting struct type.",
+            error_code=INVALID_PARAMETER_VALUE,
+        )
+
+    if (
+        isinstance(elem_type, ArrayType)
+        and isinstance(elem_type.elementType, primitive_types)
+    ):
+        # 2D array
+        return
+
+    raise MlflowException(
+        f"'spark_udf' return type does not support array type {array_type}, "
```

Comment:
`does not support array type` might sound like `does not support any array type`